### DISCUSSION TITLE
How to disable install-tool when using binarySource=docker

### DISCUSSION THREAD

**Discussion by not7cd** [renovatebot/renovate#26966] [Request Help]
### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us what version of Renovate you run.

1.37.162

### If you're self-hosting Renovate, select which platform you are using.

github.com

### What is your question?

I need to run `pip-compile` within an environment that has all system dependencies required by Python dependencies. The easiest way is to use a Docker image. I was able to do it by setting `binarySource` to `docker` as mentioned by the docs and setting `sidecarImage` to one from our registry. To my surprise, Renovate expects `install-tool` to be present in the sidecar image. Can I disable this? I see one hack to add a dummy script to my images that acts as `install-tool` and returns true.

### Logs (if relevant)

<details><summary>Logs</summary>

```

Command failed: docker run --rm --name=renovate_sidecar --label=renovate_child -v "/tmp/renovate/repos/github/my-company/my-repo":"/tmp/renovate/repos/github/my-company/my-repo" -v "/tmp/renovate/cache":"/tmp/renovate/cache" -e PIP_CACHE_DIR -e PIP_INDEX_URL -e CONTAINERBASE_CACHE_DIR -w "/tmp/renovate/repos/github/my-company/my-repo" registry.company.dev/my-base-images/image-with-system-deps:1.2.3 bash -l -c "install-tool python 3.12.1 && install-tool pip-tools 7.3.0 && pip-compile -v --no-emit-index-url requirements.in"
bash: line 1: install-tool: command not found

```

</details>

**Comment by rarkins** [comment]
binarySource=docker means:
- Run the sidecar image
- Install the necessary tool(s)
- Run the command(s)

There's no way to disable any of these steps.

Are you asking this as an end user, or as part of your development for pip-compile? Because if the latter it would be better if you can describe in full detail what you're trying to achieve. e.g. "When Renovate installs python, it is missing a few common system dependencies which I need installed"?

**Reply by not7cd** [reply]
Yea, this is an end user perspective. I guess I'm falling into something that [documentation for binary source mentions](https://docs.renovatebot.com/self-hosted-configuration/#binarysource).
> If you have a use case where you cannot use binarySource=install but can use binarySource=docker then please share it in a GitHub Discussion so that the maintainers can understand it.

Or even outside it. Because I don't really care for `pip-compile` version, as it is already preinstalled in the provided Docker image. I care for system packages that are required for Python packages to install. This is often the case for packages that provide bindings, ex.: [`pygdal`](https://pypi.org/project/pygdal/) needs Debian package `libgdal1-dev`. 

Furthermore, I have discovered that Renovate will always try to pull sidecar image defined in configuration:
https://github.com/renovatebot/renovate/blob/4ab0d9b4fbbdc462c0ac371446ff25b6784e7b4d/lib/util/exec/docker/index.ts#L270
This is not desired because I'm using a locally built image. We are not pushing dev images into the registry, except the base one. For my PoC I commented out this line as I'm running Renovate in a CI with these steps:
1. clone repository
2. `docker build --target=dev --tag=custom_sidecar repository/Dockerfile`
3. `renovate --binary-source=docker --docker-sidecar-image=custom_sidecar`

I'm basically saying to Renovate: do your job, but in an environment provided by me because I know what tools and system deps I want.
This is needed, because `pip-compile` command will always fail, and only a partial PR will be created.

**Reply by rarkins** [reply]
We meant "If binarySource=docker satisfies your needs, when binarySource=install does not, then let us know". So it wasn't meant as "If binarySource=docker does not fully satisfy your needs then let us know". We wish to deprecate binarySource=docker.

Is there anything preventing you running Renovate in a custom image which you've extended to include the tools you need?

Alternatively to overcome the immediate problem can you make `install-tool` a no-op executable in your path?

**Reply by not7cd** [reply]
> running Renovate in a custom image

Not knowing which system dependencies are needed. Every project may need different ones. I was coming from an angle where Renovate executes for every detected repository. So running an appropriate image for each repo seems like a way to go. At first, I was experimenting with a `CUSTOM_COMPILE_COMMAND` and then with `dockerSidecarImage`. But I guess this relation could be switched. 

I would imagine installing Renovate in every repository dedicated dev image with `pip-compile` preinstalled. So custom Renovate image per project. It would be the same for most, but we really care for cases where system dependencies are needed. That would then require supplying envs to that image somehow. If I understand correctly, I would set `--binarySource=global` then.
